File: 03_Stack/02_Min Stack.java
class MinStack {

    List<Integer> st;
    List<Integer> minVals;

    public MinStack() {
        st = new ArrayList<>();
        minVals = new ArrayList<>();
    }

    public void push(int val) {
        st.add(val);
        if (minVals.isEmpty() || val <= minVals.get(minVals.size() - 1))
            minVals.add(val);
    }

    public void pop() {
        int top = st.remove(st.size() - 1);
        if (minVals.get(minVals.size() - 1) == top)
            minVals.remove(minVals.size() - 1);
    }

    public int top() {
        return st.get(st.size() - 1);
    }

    public int getMin() {
        return minVals.get(minVals.size() - 1);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 03_Stack/02_Min Stack.py
class MinStack(object):

    def __init__(self):
        self.st = []
        self.minVals = []

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """

        self.st.append(val)
        if len(self.minVals) == 0 or val <= self.minVals[-1]:
            self.minVals.append(val)
        

    def pop(self):
        """
        :rtype: None
        """
        if self.minVals[-1] == self.st[-1]:
            self.minVals.pop()
        self.st.pop()

    def top(self):
        """
        :rtype: int
        """
        return self.st[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self.minVals[-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()